core/unit_test/TestGraph.hpp

```diff
@@ -349,14 +340,207 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), zero_work_reduce) {
+                         std::integral_constant<size_t, TargetIndex>,
+                         typename ViewType::non_const_value_type)
+    -> FetchValuesAndContribute<ViewType, TargetIndex, NumIndices>;
+
+// Ensure that we can handle the simple diamond use case.
+//
+// topology     stream-based approach       graph-based
+//
+//   A          A(exec_0)                   Using the API to add nodes, no
+//  / \         fence(exec_0)               user-facing fence anymore because
+// B   C        B(exec_0)   C(exec_1)       we'd like to rely on the graph to
+//  \ /         fence(exec_1)               enforce dependencies.
+//   D          D(exec_0)
+TEST_F(TEST_CATEGORY_FIXTURE(graph), diamond) {
+  const auto execution_space_instances =
+      Kokkos::Experimental::partition_space(ex, 1, 1, 1, 1);
+
+  const auto exec_0 = execution_space_instances.at(0);
+  const auto exec_1 = execution_space_instances.at(1);
+  const auto exec_2 = execution_space_instances.at(2);
+  const auto exec_3 = execution_space_instances.at(3);
+
+  using policy_t = Kokkos::RangePolicy<TEST_EXECSPACE>;
+  using view_t   = Kokkos::View<int*, TEST_EXECSPACE>;
+  using view_h_t = Kokkos::View<int*, Kokkos::HostSpace>;
+
+  view_t data(Kokkos::view_alloc(ex, "diamond - data"), 4);
+
+  constexpr int value_A = 42, value_B = 27, value_C = 13, value_D = 147;
+  std::integral_constant<size_t, 0> index_A;
+  std::integral_constant<size_t, 1> index_B;
+  std::integral_constant<size_t, 2> index_C;
+  std::integral_constant<size_t, 3> index_D;
+
+  auto graph = Kokkos::Experimental::create_graph(exec_2, [&](auto root) {
+    auto node_A = root.then_parallel_for(
+        policy_t(exec_0, 0, 1),
+        FetchValuesAndContribute(data, index_A, value_A));
+
+    auto node_B = node_A.then_parallel_for(
+        policy_t(exec_0, 0, 1),
+        FetchValuesAndContribute(data, {index_A}, index_B, value_B));
+    auto node_C = node_A.then_parallel_for(
+        policy_t(exec_1, 0, 1),
+        FetchValuesAndContribute(data, {index_A}, index_C, value_C));
+
+    auto node_D =
+        Kokkos::Experimental::when_all(node_B, node_C)
+            .then_parallel_for(policy_t(exec_0, 0, 1),
+                               FetchValuesAndContribute(
+                                   data, {index_B, index_C}, index_D, value_D));
+  });
+  graph.instantiate();
+
+  // TODO Check that kernels are running on the execution space instance of
+  //      their policy if the defaulted graph implementation is used.
+  graph.submit(exec_3);
+
+  view_h_t data_host(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "diamond - data - host"),
+      4);
+  Kokkos::deep_copy(exec_3, data_host, data);
+
+  exec_3.fence();
+
+  ASSERT_EQ(data_host(index_A()), value_A);
+  ASSERT_EQ(data_host(index_B()), value_A + value_B);
+  ASSERT_EQ(data_host(index_C()), value_A + value_C);
+  ASSERT_EQ(data_host(index_D()), 2 * value_A + value_B + value_C + value_D);
```

Comment:
```suggestion
  ASSERT_EQ(data_host(index_A), value_A);
  ASSERT_EQ(data_host(index_B), value_A + value_B);
  ASSERT_EQ(data_host(index_C), value_A + value_C);
  ASSERT_EQ(data_host(index_D), 2 * value_A + value_B + value_C + value_D);
```

Reply:
I know it's ugly... I was disappointed too.

But your proposal does not compile. Under the `clang`+`HPX` configuration used in the CI, you'd get:
```bash
In file included from /workspaces/kokkos/core/src/Kokkos_View.hpp:25:
/workspaces/kokkos/core/src/View/Kokkos_ViewLegacy.hpp:493:5: error: static_assert failed due to requirement 'Kokkos::Impl::are_integral<std::integral_constant<unsigned long, 5>>::value'
    static_assert(Kokkos::Impl::are_integral<Is...>::value);
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspaces/kokkos/core/src/View/Kokkos_ViewLegacy.hpp:506:5: note: in instantiation of function template specialization 'Kokkos::View<int *, Kokkos::HostSpace>::check_operator_parens_valid_args<std::integral_constant<unsigned long, 5>>' requested here
    check_operator_parens_valid_args(i0);
    ^
/workspaces/kokkos/core/unit_test/TestGraph.hpp:544:22: note: in instantiation of function template specialization 'Kokkos::View<int *, Kokkos::HostSpace>::operator()<std::integral_constant<unsigned long, 5>>' requested here
  ASSERT_EQ(data_host(index_F),
                     ^
1 error generated.
```

Reply:
Ugggg
cc @crtrott

Reply:
I guess you'd need some "is convertible to" instead of "is integral".

Reply:
@dalg24 @crtrott Do we just ignore this, and expect the next `View` implementation to allow for this ?

Reply:
I would say add a note as a comment would not be a bad idea but not blocking

Reply:
Should I had the comment on every line for which it would be nice to write `data(index_...)` instead of `data(index_...())` ? (*i.e.* how to make the comment visible, not repeated a 100x times)